### ISSUE TITLE
AbstractChannelHandlerContext CPU 100%

### ISSUE THREAD

**Issue by Lewik** [reactor/reactor-netty#381]
Netty used in spring integration application.
After some time cpu goes to 100% and application stop handling messages from queue (amqp and stomp)

Application inside docker

jstack of "hanged" thread
```
"tcp-client-loop-nio-4" #25 daemon prio=5 os_prio=0 tid=0x00007fc90835b000 nid=0x24 runnable [0x00007fc94d3af000]
   java.lang.Thread.State: RUNNABLE
	at io.netty.channel.AbstractChannelHandlerContext.findContextOutbound(AbstractChannelHandlerContext.java:946)
	at io.netty.channel.AbstractChannelHandlerContext.flush(AbstractChannelHandlerContext.java:746)
	at reactor.ipc.netty.channel.ChannelOperationsHandler.drain(ChannelOperationsHandler.java:415)
	at reactor.ipc.netty.channel.ChannelOperationsHandler.channelWritabilityChanged(ChannelOperationsHandler.java:172)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelWritabilityChanged(AbstractChannelHandlerContext.java:434)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelWritabilityChanged(AbstractChannelHandlerContext.java:416)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelWritabilityChanged(AbstractChannelHandlerContext.java:409)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelWritabilityChanged(ChannelInboundHandlerAdapter.java:119)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelWritabilityChanged(AbstractChannelHandlerContext.java:434)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelWritabilityChanged(AbstractChannelHandlerContext.java:416)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelWritabilityChanged(AbstractChannelHandlerContext.java:409)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelWritabilityChanged(ChannelInboundHandlerAdapter.java:119)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelWritabilityChanged(AbstractChannelHandlerContext.java:434)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelWritabilityChanged(AbstractChannelHandlerContext.java:416)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelWritabilityChanged(AbstractChannelHandlerContext.java:409)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelWritabilityChanged(ChannelInboundHandlerAdapter.java:119)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelWritabilityChanged(AbstractChannelHandlerContext.java:434)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelWritabilityChanged(AbstractChannelHandlerContext.java:416)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelWritabilityChanged(AbstractChannelHandlerContext.java:409)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelWritabilityChanged(DefaultChannelPipeline.java:1457)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelWritabilityChanged(AbstractChannelHandlerContext.java:434)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelWritabilityChanged(AbstractChannelHandlerContext.java:416)
	at io.netty.channel.DefaultChannelPipeline.fireChannelWritabilityChanged(DefaultChannelPipeline.java:977)
	at io.netty.channel.ChannelOutboundBuffer.fireChannelWritabilityChanged(ChannelOutboundBuffer.java:607)
	at io.netty.channel.ChannelOutboundBuffer.setWritable(ChannelOutboundBuffer.java:573)
	at io.netty.channel.ChannelOutboundBuffer.decrementPendingOutboundBytes(ChannelOutboundBuffer.java:194)
	at io.netty.channel.ChannelOutboundBuffer.remove(ChannelOutboundBuffer.java:259)
	at io.netty.channel.ChannelOutboundBuffer.removeBytes(ChannelOutboundBuffer.java:338)
	at io.netty.channel.socket.nio.NioSocketChannel.doWrite(NioSocketChannel.java:409)
	at io.netty.channel.AbstractChannel$AbstractUnsafe.flush0(AbstractChannel.java:934)
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.forceFlush(AbstractNioChannel.java:367)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:641)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:582)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:499)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:461)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:884)
	at java.lang.Thread.run(Thread.java:748)
```

### Netty version
4.1.25
### JVM version (e.g. `java -version`)
openjdk version "1.8.0_171"
OpenJDK Runtime Environment (build 1.8.0_171-8u171-b11-1~deb9u1-b11)
OpenJDK 64-Bit Server VM (build 25.171-b11, mixed mode)

### OS version (e.g. `uname -a`)
Linux 66fb569f18f2 3.16.0-4-amd64 #1 SMP Debian 3.16.43-2 (2017-04-30) x86_64 GNU/Linux

I am not shure that this is netty.
jars in docker with "netty":
netty-buffer-4.1.25.Final.jar
netty-codec-4.1.25.Final.jar
netty-codec-http-4.1.25.Final.jar
netty-codec-socks-4.1.25.Final.jar
netty-common-4.1.25.Final.jar
netty-handler-4.1.25.Final.jar
netty-handler-proxy-4.1.25.Final.jar
netty-resolver-4.1.25.Final.jar
netty-transport-4.1.25.Final.jar
netty-transport-native-epoll-4.1.25.Final-linux-x86_64.jar
netty-transport-native-unix-common-4.1.25.Final.jar
reactor-netty-0.7.8.RELEASE.jar

**Comment by violetagg** [comment]
Hi,

Any test scenario that we can use in order to reproduce the issue?

Regards,
Violeta

**Comment by Lewik** [comment]
@violetagg This freeze appears suddenly. I don't know steps to reproduce, I even don't know when exactly in freezes. It can be 30 minutes after start, or 3 hours.

**Comment by Lewik** [comment]
I see two threads "tcp-client-loop-nio-4" in jstack. nio-1, 2, 3 - they are doubled.
I see SingleThreadEventExecutor in the stack beginning.

Is that duplicates OK?

**Comment by violetagg** [comment]
Can you attach the output from `jstack`?
Also at least if you can describe the scenario that you execute.

**Comment by Lewik** [comment]
Additional info:
1. @normanmaurer think that this "This looks more like a project reactor bug that will consume a lot of cpu while try to drain some queue..."
https://github.com/netty/netty/issues/8092

2. This bug appears with Spring 5. But with Spring 4 all OK.

**Comment by yu199195** [comment]
I use webflux websocket ， It's a problem，
this is jstack：
"bbex-webflux-http-nio-141" #166 prio=5 os_prio=0 tid=0x00007f226be6a000 nid=0x7ba9 runnable [0x00007f2139493000]
      java.lang.Thread.State: RU: RUNNABLE
	at at io.netty.channel.AbstractChannelHandlerContext.findContextOuxtOutbound(nd(AbstractChannelHandlerContext.java:946)
	a)
	at at io.netty.channel.AbstractChannelHandlerContext.flush(Abs(sh(AbstractChannelHandlerContext.java:746)
	a)
	at at reactor.ipc.netty.channel.ChannelOperationsHandler.drain(Cha(in(ChannelOperationsHandler.java:415)
	a)
	at at reactor.ipc.netty.channel.ChannelOperationsHandler.flush(Cha(sh(ChannelOperationsHandler.java:192)
	a)
	at at io.netty.channel.AbstractChannelHandlerContext.invokeFlush0(sh0(h0(AbstractChannelHandlerContext.java:776)
	a)
	at at io.netty.channel.AbstractChannelHandlerContext.invokeFlush(Ash(sh(AbstractChannelHandlerContext.java:768)
	a)
	at at io.netty.channel.AbstractChannelHandlerContext.access$150$1500(00(AbstractChannelHandlerContext.java:38)
	a)
	at at io.netty.channel.AbstractChannhannelHandlerContext$xt$WriteAndFlushTask.write(Abs(te(AbstractChannelHandlerContext.java:1129)
	a)
	at at io.netty.channel.AbstractChannhannelHandlerContext$xt$AbstractWriteTask.run(Abs(un(AbstractChannelHandlerContext.java:1070)
	a)
	at at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(Ate(te(AbstractEventExecutor.java:163)
	a)
	at at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(Sks(ks(SingleThreadEventExecutor.java:404)
	a)
	at at io.netty.channel.nio.NioEventLoop.run(Nio(un(NioEventLoop.java:464)
	a)
	at at io.netty.util.concurrent.SingleThreadEeadEventExecutor$or$5.run(Sin(un(SingleThreadEventExecutor.java:884)
	a)
	at at java.lang.Thread.run(Thr(un(Thread.java:748))

**Comment by yu199195** [comment]
Is this problem solved？

**Comment by violetagg** [comment]
@yu199195 Nope it is not solved. Thanks for the demo project. We are looking at it.

**Comment by yu199195** [comment]
[calvin-websocket.tar.gz](https://github.com/reactor/reactor-netty/files/2378805/calvin-websocket.tar.gz)
@violetagg

**Comment by violetagg** [comment]
@yu199195 Based on your sources I tried to extract some scenario in order to reproduce the issue, unfortunately without success. Can you explain what the setup is and how to run your example? Can you try reproduce the issue with the current SNAPSHOT version?

**Comment by Lewik** [comment]
I just tried update to latest spring version

org.springframework.boot:spring-boot:2.1.0.RELEASE
org.springframework:spring-core:5.1.2.RELEASE
io.projectreactor.netty:reactor-netty:0.8.2.RELEASE

and

org.springframework.boot:spring-boot:2.1.1.BUILD-SNAPSHOT
org.springframework:spring-core:5.1.3.BUILD-SNAPSHOT
io.projectreactor.netty:reactor-netty:0.8.3.BUILD-SNAPSHOT

Same behavior.

On KVM virtual machine it hangs.
On my notebook under VirtualBox - not hangs (may be it just more powerfull)

@smaldini @violetagg 
This issue blocks moving our project to Spring5. If you have time, I will appreciate you to any help, I could give access to server. lllewik@gmail.com

**Comment by Lewik** [comment]
`io.projectreactor.netty:reactor-netty:0.8.4.RELEASE`
No hangs within 1 day.

@yu199195 Could you check?

**Comment by yu199195** [comment]
@Lewik  yes，We used websocket in production，This happens after a period of running
So，We can only use native netty。
Webflux，Not mature enough， at least for now。

**Comment by Lewik** [comment]
@yu199195 Check with new version, please.

**Comment by Lewik** [comment]
There are no freezes with 0.8.4.RELEASE within about week. There was freezes with previous versions within 1-2 hours. As I (nobody) don't know why freezes happens and how to check this: Personally I will assume that freezes was fixed in 0.8.4.RELEASE. If anybody will face freezes again - please text here. Thanks @violetagg for her patience.

**Comment by violetagg** [comment]
Let's close this. If the issue appears again we will reopen it.
